(#247) new formatter fix bug that comment would be eliminated

YanHerChen opened the pull request.

YanHerChen committed:
new-formatted-elz

dannypsnl commented on src/codegen/formatted_elz_new/tests.rs:
```diff
@@ -0,0 +1,227 @@
+}
+
+// helpers, must put tests before this line
+fn parse_code(code: &str) -> String {
```
This is a bad name, I think ”format code” would be better.

YanHerChen replied:
I agree.

dannypsnl commented on src/codegen/formatted_elz_new/tests.rs:
```diff
@@ -0,0 +1,227 @@
+fn class_multiple_super_types() {
+    let program = parse_code("class Foo<:Bar,Tool{}");
+    assert_eq!(
+        program,
```
Suggested name: ”formatted code”

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+#[cfg(test)]
```
Remember remove the old module

YanHerChen commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+mod tests;
+
+fn formatted_elz(code: String) -> String {
+    let mut indent = 0i32;
```
### Handling indent

count integer to record the indent, e.q.
```
class foo{
    fn function1(){
    }
}
```
indent is 2.

YanHerChen commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+
+fn formatted_elz(code: String) -> String {
+    let mut indent = 0i32;
+    let mut comment_line = false;
```
if read comment line, then comment_line is true

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+        }
+        i += 1;
+    }
+    println!("{}", s);
```
Don't print out the result! This is not expected behavior. Also, not a good debug habit. Try debugger in future.

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+                }
+                //delete double blank
+                "}" => {
+                    indent -= 1;
```
Consider the ”state function” trick.

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+            }
+        } else {
+            match code_to_char[i] {
+                //{ } ; need to add newline
```
Comment should keep one space like

// ...

Not

//...

YanHerChen commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+fn formatted_elz(code: String) -> String {
+    let mut indent = 0i32;
+    let mut comment_line = false;
+    let mut semicolon_symbol = false;
```
if read `;` , then semicolon_symbol is true, the symbol `;` means end of line, and need to set newline.

YanHerChen commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+    let mut indent = 0i32;
+    let mut comment_line = false;
+    let mut semicolon_symbol = false;
+    let mut past_symbol = false;
```
past_symbol handling the situation like:
```
Foo = Foo{x: 1};
```
symbol `{` should not newline in this situation, when formatter read `=` before `{`, then past_symbol is true

YanHerChen commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+    let mut semicolon_symbol = false;
+    let mut past_symbol = false;
+    let mut s = String::from("");
+    let clear_multiple_blank = code.replace("  ", "").replace("\n", "\\n");
```
Formatter replace double blank in first time, to clear multi-blank.

dannypsnl replied:
Have to notice this would take 2n to deal this, n is code size.

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+    s
+}
+
+fn add_tab(tab: i32) -> String {
```
Add indent?

YanHerChen commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+                i += 1;
+            } else{                
+                s.push_str(code_to_char[i]);
+            }
```
if read comment line, formatter will ignore until meet `\n`.

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,184 @@
+}
+
+fn check_symbol_behind_space(symbol: &str) -> bool {
+    let mut b = false;
```
Don't do this, return value directly.

For example

if symbol == a { true }
if...
if...
else { false }

YanHerChen committed:
new formatter fix bug that comment would be eliminated

dannypsnl commented on src/codegen/formatted_elz_new/tests.rs:
```diff
@@ -0,0 +1,227 @@
+use super::formatted_elz;
```
How about ”use super::*;” to avoid any changes about this line?

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,183 @@
+mod tests;
```
I say remove old formatter module, not metadata for rust compiler. Crate test attr helps rust compiler know mod tests is testing module. Please at least run cargo test before push any commits.

dannypsnl commented on src/codegen/formatted_elz_new/mod.rs:
```diff
@@ -0,0 +1,183 @@
+    let code_to_char: Vec<&str> = clear_multiple_blank.split("").collect();
+
+    let mut i = 0;
+    while i < code_to_char.len() {
```
Try .iter().enumerate()

dannypsnl commented on src/codegen/formatted_elz_new/tests.rs:
```diff
@@ -0,0 +1,227 @@
+fn format_code(code: &str) -> String {
+    let s = formatted_elz(code.to_string());
+    s
+}
```
EOF, please ensure you're using formatter in your developing environment.

dannypsnl commented on src/codegen/formatted_elz_new/tests.rs:
```diff
@@ -0,0 +1,227 @@
+}
+
+#[test]
+fn function_call_in_function() {
```
I don't think we need this. But can test ”if” in function to ensure nested block would work.
Also can test comment in function.

YanHerChen committed:
(#247) remove old formatter module

dannypsnl commented on src/codegen/formatted_elz/mod.rs:
```diff
@@ -1,354 +1,188 @@
-                s.push_str("}");
-                s
-            }
+fn add_indent(tab: i32) -> String {
```
Not tab, how about level?

dannypsnl commented on src/codegen/formatted_elz/mod.rs:
```diff
@@ -1,354 +1,188 @@
-                s
+        } else {
+            match code_to_char[i] {
+                //{ } ; need to add newline
```
// { }

NOT

//{ }

dannypsnl replied:
Please fix all comments like that

dannypsnl commented on src/codegen/formatted_elz/mod.rs:
```diff
@@ -1,354 +1,188 @@
-        }
+fn check_symbol_behind_space(symbol: &str) -> bool {
+    if symbol == ";" {
+        return true;
```
In Rust, an exhaust if-else block doesn't need a return keyword.

You can use {true} to get same result.

dannypsnl commented on src/codegen/formatted_elz/tests.rs:
```diff
@@ -122,19 +123,102 @@ fn nested_block() {
-    };
-    FormatTopAstList(program)
+fn format_code(code: &str) -> String {
+    let s = formatted_elz(code.to_string());
```
I don't think we need to wrap ”formatted_elz.”

Directly call it would be better.

And why not name it ”format_code”? Notice the original name is because it belongs to AST. And the mod also needs a new name, how about ”formatter”?

YanHerChen committed:
(#247) rename function
1. name formatter replace formatted_elz in mod.rs
2. fix `//comment` to `// comment` in mod.rs
3. delete fn format_code, and directly call formatter in tests.rs

dannypsnl commented on src/cmd/fmt.rs:
```diff
@@ -1,4 +1,4 @@
-use crate::codegen::formatted_elz::formatted_elz;
+use crate::codegen::formatted_elz::formatter;
```
You didn't read it correctly. Rename mod to formatter, not function. Rename function to format_elz.

dannypsnl commented on src/cmd/fmt.rs:
```diff
@@ -34,8 +33,7 @@ fn handle_dir(dir: &Path) -> Result<(), Box<dyn std::error::Error>> {
         let code = std::fs::read_to_string(file_path)?;
-        let program = Parser::parse_program(file_path.to_string_lossy().to_string(), code)?;
-        let result = FormatTopAstList(program).formatted_elz(DEFAULT_LEVEL);
+        let result = formatter(code.to_string());
```
Don't you feel formatter as function name is weird?

YanHerChen committed:
(#247) rename
rename mod name formatted_elz to formatter
rename funrction name formatter to format_elz
rename parameter tab to level in fn add_indent

dannypsnl committed:
Merge branch 'master' into master
